### ISSUE TITLE
Slow DNS lookup through Netgear DG834G V4 router

### ISSUE THREAD

**Issue by martinpitt** [systemd/systemd#3980]
### Submission type
- [X] Bug report
- [ ] Request for enhancement (RFE)
### systemd version the issue has been seen with

> 231

_NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!_
### Used distribution

> Ubuntu yakkety

With this router, lookups via `nss-resolve` take very long (with glibc it's fast). I asked the original reporter to get a `resolved` output with debug logging enabled while running `systemd-resolve www.google.com`: https://launchpadlibrarian.net/279608636/debug.txt

This shows a couple of timeouts. I will now ask the reporter to re-try with disabling DNSSEC, for comparison.

(Downstream bug: https://launchpad.net/bugs/1609546)

**Comment by martinpitt** [comment]
The original reporter confirms that `DNSSEC=no` avoids the hangs.

**Comment by poettering** [comment]
So, looking at the provided logs, it appears as if the server doesn't reply at all on any of our fancier requests. We then gradually downgrade until we get a reply, but that takes a while, as we wait for repeated timeouts. Not sure what we can do about this, really. The first lookup should be slow, but subsequent lookups should be much faster then, as we already downgraded. Any chance you could verify if that works correctly, i.e. that after the first, slow lookup, things get fast again?

**Comment by colinl** [comment]
Something seems to have changed subtly since I (the downstream bug reporter)originally reported this. On Ubuntu Zesty Beta I now see a timeout rather than a long delay. I have attached a new log in case it is helpful. I see that it takes 3 to 5 seconds for each query and the whole thing times out after 30 seconds.
With DNSSEC disabled it works ok.
[debug.txt](https://github.com/systemd/systemd/files/868617/debug.txt)

**Comment by quozl** [comment]
On Ubuntu Zesty Beta I see a timeout as well, but different router, a Huawei E5573S, which drops the `SOA .` query from `systemd-resolved` but is happy to answer `dig @192.168.1.1 -t SOA .`  With DNSSEC disabled `systemd-resolved` works fine; the `SOA .` query is absent.  Hope these may help:
* the debug log, [debug.txt](https://github.com/systemd/systemd/files/912028/debug.txt);
* wireshark dissection of working query made by `dig(1)`, [query-that-works.txt](https://github.com/systemd/systemd/files/912029/query-that-works.txt);
* wireshark dissection of dropped query made by `systemd-resolved`, [query-that-fails.txt](https://github.com/systemd/systemd/files/912030/query-that-fails.txt); and,
* the difference in queries is in the _Additional RRs_ of type _OPT_.

Admittedly, possibly a different problem than Colin's downstream reporter, but the way the peer simply drops the query is very familiar.

**Comment by Phlamethrower** [comment]
I'm seeing similar issues to those described here with systemd-resolve 232 on a Lubuntu 17.04 install, with a DG384G v3 router.

When the machine boots, systemd-resolve degrades to the UDP feature set, and all is well. But when it switches back to the full UDP+EDNS0+DO+LARGE feature set, for some domains it will get corrupt replies from the router. And these corrupt replies don't appear to trigger the feature set degradation logic, leaving those domains in a permanently-unresolvable state. I do have DNSSEC disabled, but that doesn't seem to be enough for this router.

* wireshark dissection of working query made by `systemd-resolve debian.org`: [works-systemd-resolve.txt](https://github.com/systemd/systemd/files/959753/works-systemd-resolve.txt)
* wireshark dissection of working query made by `dig +noadflag debian.org @10.0.0.1`: [works-dig.txt](https://github.com/systemd/systemd/files/959757/works-dig.txt)
* wireshark dissection of broken query and response for `systemd-resolve debian.org`: [broken-systemd-resolve.txt](https://github.com/systemd/systemd/files/959758/broken-systemd-resolve.txt)
* `systemd-resolve` debug log for the broken query: [debug.txt](https://github.com/systemd/systemd/files/959759/debug.txt)

**Comment by colinl** [comment]
@Phlamethrower that sounds like a different bug to me. Also do you mean a DG384 or DG834?

**Comment by poettering** [comment]
@Phlamethrower your issue appears unrelated, please do not hijack bugs. Could you also try with resolved 233 first? resolved received a larger numbre of changes in 233, and there's a good chance your specific problem is already fixed with it.

**Comment by Phlamethrower** [comment]
@poettering Sure, I'll keep an eye out for when a newer release comes through a stable Ubuntu channel, and then open a new issue if necessary. I've already spent far too many hours working through issues with 17.04 to want to try updating to an unstable version!
@colinl DG834G, sorry.

**Comment by perolausson** [comment]
I am the happy owner of a similar HW device (technicolor TG589vn) with an unruly DNS server and I think the problem is that it starts out working well by downgrading to UDP works fine, but then it upgrades and for whatever reason it get's stuck at being upgraded rendering the system unusable. So I am not so sure this is an unrelated / different problem.

I am currently trying out the upstream master build to see if my hairy problems can be resolved.

I believe the outcome is a fail.

Recap: when it was happy @ 232:

```
Sep 29 10:39:53 janus systemd[1]: Started Network Name Resolution.
Sep 29 10:42:34 janus systemd-resolved[8365]: Switching to system DNS server 192.168.1.1.
Sep 29 10:42:34 janus systemd-resolved[8365]: Switching to DNS server 192.168.1.1 for interface eno1.
Sep 29 10:42:35 janus systemd-resolved[8365]: Using degraded feature set (UDP) for DNS server 192.168.1.1.
Sep 29 10:42:35 janus systemd-resolved[8365]: Using degraded feature set (UDP) for DNS server 192.168.1.1.
...
Sep 29 10:57:20 janus systemd-resolved[8961]: Processing incoming packet on transaction 30864.
Sep 29 10:57:20 janus systemd-resolved[8961]: OPT RR contained RFC6975 data, ignoring.
Sep 29 10:57:20 janus systemd-resolved[8961]: Server doesn't support EDNS(0) properly, downgrading feature level...
Sep 29 10:57:20 janus systemd-resolved[8961]: Using degraded feature set (UDP) for DNS server 192.168.1.1.
Sep 29 10:57:20 janus systemd-resolved[8961]: Server feature level is now lower than when we began our transaction. Restarting with new ID.
Sep 29 10:57:20 janus systemd-resolved[8961]: Transaction 30864 is now 45709.
Sep 29 10:57:20 janus systemd-resolved[8961]: Cache miss for zeus2.lan IN AAAA
Sep 29 10:57:20 janus systemd-resolved[8961]: Transaction 45709 for <zeus2.lan IN AAAA> scope dns on */*.
Sep 29 10:57:20 janus systemd-resolved[8961]: Using feature level UDP for transaction 45709.
Sep 29 10:57:20 janus systemd-resolved[8961]: Sending query packet with id 45709.
Sep 29 10:57:20 janus systemd-resolved[8961]: Processing incoming packet on transaction 23063.
Sep 29 10:57:20 janus systemd-resolved[8961]: OPT RR contained RFC6975 data, ignoring.
Sep 29 10:57:20 janus systemd-resolved[8961]: Verified we get a response at feature level UDP from DNS server 192.168.1.1.
Sep 29 10:57:20 janus systemd-resolved[8961]: Server doesn't support EDNS(0) properly, downgrading feature level...
Sep 29 10:57:20 janus systemd-resolved[8961]: Using degraded feature set (UDP) for DNS server 192.168.1.1.
Sep 29 10:57:20 janus systemd-resolved[8961]: Server feature level is now lower than when we began our transaction. Restarting with new ID.
Sep 29 10:57:20 janus systemd-resolved[8961]: Transaction 23063 is now 24250.
Sep 29 10:57:20 janus systemd-resolved[8961]: Cache miss for zeus2.lan IN AAAA
Sep 29 10:57:20 janus systemd-resolved[8961]: Transaction 24250 for <zeus2.lan IN AAAA> scope dns on eno1/*.
Sep 29 10:57:20 janus systemd-resolved[8961]: Using feature level UDP for transaction 24250.
Sep 29 10:57:20 janus systemd-resolved[8961]: Sending query packet with id 24250.
Sep 29 10:57:20 janus systemd-resolved[8961]: Processing incoming packet on transaction 45709.
Sep 29 10:57:20 janus systemd-resolved[8961]: Not caching negative entry without a SOA record: zeus2.lan IN AAAA
Sep 29 10:57:20 janus systemd-resolved[8961]: Transaction 45709 for <zeus2.lan IN AAAA> on scope dns on */* now complete with <success> from network (unsigned).
Sep 29 10:57:20 janus systemd-resolved[8961]: Freeing transaction 24250.
Sep 29 10:57:20 janus systemd-resolved[8961]: Sending response packet with id 58849 on interface 1/AF_INET.
Sep 29 10:57:20 janus systemd-resolved[8961]: Freeing transaction 45709.
Sep 29 10:57:20 janus systemd-resolved[8961]: Got DNS stub UDP query packet for id 18270
Sep 29 10:57:20 janus systemd-resolved[8961]: Looking up RR for zeus2.lan IN A.
Sep 29 10:57:20 janus systemd-resolved[8961]: Cache miss for zeus2.lan IN A
Sep 29 10:57:20 janus systemd-resolved[8961]: Transaction 43344 for <zeus2.lan IN A> scope dns on */*.
Sep 29 10:57:20 janus systemd-resolved[8961]: Using feature level UDP for transaction 43344.
Sep 29 10:57:20 janus systemd-resolved[8961]: Using DNS server 192.168.1.1 for transaction 43344.
Sep 29 10:57:20 janus systemd-resolved[8961]: Sending query packet with id 43344.
Sep 29 10:57:20 janus systemd-resolved[8961]: Cache miss for zeus2.lan IN A
Sep 29 10:57:20 janus systemd-resolved[8961]: Transaction 23032 for <zeus2.lan IN A> scope dns on eno1/*.
Sep 29 10:57:20 janus systemd-resolved[8961]: Using feature level UDP for transaction 23032.
Sep 29 10:57:20 janus systemd-resolved[8961]: Using DNS server 192.168.1.1 for transaction 23032.
Sep 29 10:57:20 janus systemd-resolved[8961]: Sending query packet with id 23032.
Sep 29 10:57:20 janus systemd-resolved[8961]: Processing query...
Sep 29 10:57:20 janus systemd-resolved[8961]: Got DNS stub UDP query packet for id 46824
Sep 29 10:57:20 janus systemd-resolved[8961]: Looking up RR for zeus2.lan IN AAAA.
Sep 29 10:57:20 janus systemd-resolved[8961]: Cache miss for zeus2.lan IN AAAA
Sep 29 10:57:20 janus systemd-resolved[8961]: Transaction 43728 for <zeus2.lan IN AAAA> scope dns on */*.
Sep 29 10:57:20 janus systemd-resolved[8961]: Using feature level UDP for transaction 43728.
Sep 29 10:57:20 janus systemd-resolved[8961]: Using DNS server 192.168.1.1 for transaction 43728.
Sep 29 10:57:20 janus systemd-resolved[8961]: Sending query packet with id 43728.
Sep 29 10:57:20 janus systemd-resolved[8961]: Cache miss for zeus2.lan IN AAAA
Sep 29 10:57:20 janus systemd-resolved[8961]: Transaction 24538 for <zeus2.lan IN AAAA> scope dns on eno1/*.
Sep 29 10:57:20 janus systemd-resolved[8961]: Using feature level UDP for transaction 24538.
Sep 29 10:57:20 janus systemd-resolved[8961]: Using DNS server 192.168.1.1 for transaction 24538.
Sep 29 10:57:20 janus systemd-resolved[8961]: Sending query packet with id 24538.
Sep 29 10:57:20 janus systemd-resolved[8961]: Processing query...
Sep 29 10:57:20 janus systemd-resolved[8961]: Processing incoming packet on transaction 43344.
Sep 29 10:57:20 janus systemd-resolved[8961]: Not caching zero TTL cache entry: zeus2.lan IN A
Sep 29 10:57:20 janus systemd-resolved[8961]: Transaction 43344 for <zeus2.lan IN A> on scope dns on */* now complete with <success> from network (unsigned).
```

Main events during the day were:

```
journalctl -u systemd-resolved | egrep -i 'degrade|grace|Using feature level|now complete'| egrep -i 'degrade|grace|Using feature level|zeus2'|grep -i grace
Sep 29 10:48:11 janus systemd-resolved[8580]: Grace period over, resuming full feature set (UDP+EDNS0+DO+LARGE) for DNS server fe80::c6ea:1dff:fe6c:4ee2%2.
Sep 29 11:12:44 janus systemd-resolved[8961]: Grace period over, resuming full feature set (UDP+EDNS0+DO+LARGE) for DNS server 192.168.1.1.
Sep 29 11:12:44 janus systemd-resolved[8961]: Grace period over, resuming full feature set (UDP+EDNS0+DO+LARGE) for DNS server 192.168.1.1.
Sep 29 11:36:07 janus systemd-resolved[9696]: Grace period over, resuming full feature set (UDP+EDNS0+DO+LARGE) for DNS server 192.168.1.1.
Sep 29 17:15:15 janus systemd-resolved[9696]: Grace period over, resuming full feature set (UDP+EDNS0+DO+LARGE) for DNS server 192.168.1.1.
```
Last success:
```
journalctl -u systemd-resolved | egrep -i 'degrade|grace|now complete'| egrep -i 'degrade|grace|zeus2'|grep success|tail -1
Sep 29 16:44:51 janus systemd-resolved[9696]: Transaction 31935 for <zeus2.lan IN A> on scope dns on eno1/* now complete with <success> from network (unsigned).
```
Failures on 232:

```
Sep 29 16:53:09 janus systemd-resolved[9696]: Transaction 58901 for <zeus2.lan IN A> scope dns on eno1/*.
Sep 29 16:53:09 janus systemd-resolved[9696]: Using feature level UDP+EDNS0+DO+LARGE for transaction 58901.
Sep 29 16:53:09 janus systemd-resolved[9696]: Using DNS server fe80::c6ea:1dff:fe6c:4ee2%2 for transaction 58901.
Sep 29 16:53:09 janus systemd-resolved[9696]: Sending query packet with id 58901.
Sep 29 16:53:09 janus systemd-resolved[9696]: Processing incoming packet on transaction 58901.
Sep 29 16:53:09 janus systemd-resolved[9696]: Transaction 58901 for <zeus2.lan IN A> on scope dns on eno1/* now complete with <invalid-reply> from network (unsigned).
Sep 29 16:53:09 janus systemd-resolved[9696]: Freeing transaction 58901.
```

Failure on 233:

```
Sep 29 17:33:49 janus systemd-resolved[12084]: Using degraded feature set (UDP+EDNS0+DO) for DNS server 192.168.1.1.
Sep 29 17:39:13 janus systemd-resolved[13772]: Transaction 46730 for <zeus2.lan IN A> on scope dns on eno1/* now complete with <invalid-reply> from network (unsigned).
Sep 29 17:40:38 janus systemd-resolved[6669]: Transaction 29202 for <zeus2.lan IN A> on scope dns on eno1/* now complete with <invalid-reply> from network (unsigned).
Sep 29 17:53:29 janus systemd-resolved[6669]: Transaction 45164 for <zeus2.lan IN A> on scope dns on eno1/* now complete with <invalid-reply> from network (unsigned).
$ journalctl -u systemd-resolved | egrep -i 45164
Sep 29 17:53:29 janus systemd-resolved[6669]: Transaction 45164 for <zeus2.lan IN A> scope dns on eno1/*.
Sep 29 17:53:29 janus systemd-resolved[6669]: Using feature level UDP+EDNS0 for transaction 45164.
Sep 29 17:53:29 janus systemd-resolved[6669]: Using DNS server 192.168.1.1 for transaction 45164.
Sep 29 17:53:29 janus systemd-resolved[6669]: Sending query packet with id 45164.
Sep 29 17:53:29 janus systemd-resolved[6669]: Processing incoming packet on transaction 45164. (rcode=SUCCESS)
Sep 29 17:53:29 janus systemd-resolved[6669]: Transaction 45164 for <zeus2.lan IN A> on scope dns on eno1/* now complete with <invalid-reply> from network (unsigned).
Sep 29 17:53:29 janus systemd-resolved[6669]: Freeing transaction 45164.
```

So effectively 233 for me is worse than 232, because the minimum level is UDP+EDNS0 whereas 232 tolerated UDP.

What would solve it for me would be if I could specify the maximum supported level for my device or to say "don't try to upgrade from a working solution", ie. stop this Grace period nonsense. I can assure you this device will never magically figure out EDNS!

My settings btw:

```
$ cat /etc/systemd/network/eno1.network
[Match]
Name=eno1
[Network]
DNS=192.168.1.1
DNSSECNegativeTrustAnchors=lan
DHCP=yes
Domains=lan
[DHCP]
UseDomains=true
```

**Comment by perolausson** [comment]
I thought there might be some new directives and retried on 233 using this:

```
[Match]
Name=eno1
[Network]
DNS=192.168.1.1
DNSSEC=allow-downgrade
DNSSECNegativeTrustAnchors=lan
LLMNR=false
DHCP=yes
Domains=lan
[DHCP]
UseDomains=true
```

It didn't change anything. But I discovered another interesting detail:

Dig client:
```
dig zeus2

; <<>> DiG 9.10.3-P4-Ubuntu <<>> zeus2
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 5565
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;zeus2.                         IN      A  

;; Query time: 0 msec
;; SERVER: 127.0.0.53#53(127.0.0.53)
;; WHEN: Fri Sep 29 18:19:15 BST 2017
;; MSG SIZE  rcvd: 34

# What if I query the DNS server directly?

$ dig @192.168.1.1 zeus2.lan
;; Warning: Message parser reports malformed message packet.

; <<>> DiG 9.10.3-P4-Ubuntu <<>> @192.168.1.1 zeus2.lan
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 23840
;; flags: qr aa rd ra ad; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; QUESTION SECTION:
;zeus2.lan.                     IN      A

;; ANSWER SECTION:
zeus2.lan.              0       IN      A       192.168.1.72

;; Query time: 1 msec
;; SERVER: 192.168.1.1#53(192.168.1.1)
;; WHEN: Fri Sep 29 18:19:28 BST 2017
;; MSG SIZE  rcvd: 43

# And then retry against systemd?

$ dig zeus2.lan

; <<>> DiG 9.10.3-P4-Ubuntu <<>> zeus2.lan
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 9650
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;zeus2.lan.                     IN      A

;; ANSWER SECTION:
zeus2.lan.              0       IN      A       192.168.1.72

;; Query time: 1 msec
;; SERVER: 127.0.0.53#53(127.0.0.53)
;; WHEN: Fri Sep 29 18:19:41 BST 2017
;; MSG SIZE  rcvd: 54
```
I don't think you can even begin to understand how frustrated I am with this software.

My theory: so what happens appears to be that systemd-resolve sees a UDP packet flyby and intercepts it, and then it figures out that maybe it could start listening to the DNS server again, but in a degraded form. But I am just guessing.

Log for first fail (times match):

```
Sep 29 18:19:15 janus systemd-resolved[3785]: Got DNS stub UDP query packet for id 48405
Sep 29 18:19:15 janus systemd-resolved[3785]: Looking up RR for zeus2 IN A.
Sep 29 18:19:15 janus systemd-resolved[3785]: Sending response packet with id 48405 on interface 1/AF_INET.
Sep 29 18:19:15 janus systemd-resolved[3785]: Processing query...
```

When it works again:

```
Sep 29 18:19:15 janus systemd-resolved[3785]: Got DNS stub UDP query packet for id 48405
Sep 29 18:19:15 janus systemd-resolved[3785]: Looking up RR for zeus2 IN A.
Sep 29 18:19:15 janus systemd-resolved[3785]: Sending response packet with id 48405 on interface 1/AF_INET.
Sep 29 18:19:15 janus systemd-resolved[3785]: Processing query...
Sep 29 18:19:41 janus systemd-resolved[3785]: Got DNS stub UDP query packet for id 45605
Sep 29 18:19:41 janus systemd-resolved[3785]: Looking up RR for zeus2.lan IN A.
Sep 29 18:19:41 janus systemd-resolved[3785]: Cache miss for zeus2.lan IN A
Sep 29 18:19:41 janus systemd-resolved[3785]: Transaction 13693 for <zeus2.lan IN A> scope dns on eno1/*.
Sep 29 18:19:41 janus systemd-resolved[3785]: Using feature level UDP for transaction 13693.
Sep 29 18:19:41 janus systemd-resolved[3785]: Using DNS server 192.168.1.1 for transaction 13693.
Sep 29 18:19:41 janus systemd-resolved[3785]: Sending query packet with id 13693.
Sep 29 18:19:41 janus systemd-resolved[3785]: Processing query...
Sep 29 18:19:41 janus systemd-resolved[3785]: Processing incoming packet on transaction 13693. (rcode=SUCCESS)
Sep 29 18:19:41 janus systemd-resolved[3785]: Not validating response for 13693, used server feature level does not support DNSSEC.
Sep 29 18:19:41 janus systemd-resolved[3785]: Not caching zero TTL cache entry: zeus2.lan IN A
Sep 29 18:19:41 janus systemd-resolved[3785]: Transaction 13693 for <zeus2.lan IN A> on scope dns on eno1/* now complete with <success> from network (unsigned).
```

As I wrote this up, systemd has already decided to give up on my DNS server again.

It's hellish this software. The logic you have in it is very complicated but it isn't tested properly. I suggest you buy one of these devices off ebay and try it out.

**Comment by poettering** [comment]
So, looking at the dig output this DNS server is not getting the RRCOUNT right, i.e. it claims there were more RRs in the packet than there are. Fix in #17535
